Remove more charmstore query data

ChrisMacNaughton opened the pull request:
Additionally, makes two of the fields on-demand only, as
libcharmstore nor amulet actually reference them. The API
remains unchanged but the amount of data requested from
the charmstore should be dramatically reduced.

ChrisMacNaughton committed:
Remove more charmstore query data

Additionally, makes two of the fields on-demand only, as
libcharmstore nor amulet actually reference them. The API
remains unchanged but the amount of data requested from
the charmstore should be dramatically reduced.

gnuoy commented:
LGTM. As you said @ChrisMacNaughton there is a risk of someone accessing self.raw directly but I think its a risk worth taking.

marcoceppi commented on charmstore/lib.py:
```diff
@@ -95,13 +87,14 @@ def __init__(self, id=None, api=None, timeout=None):
         self.source = None
 
-        self.files = []
+        self.files = None
```
People iterating without validating will be in trouble here

ChrisMacNaughton replied:
Agreed, I maintained functions to be backwards compatible but didn't consider the fields carefully enough, resolved in the latest commit.

ChrisMacNaughton committed:
revert change to public field

Instead, add a new field to track if we have previously fetched
files that are part of the Entity

johnsca commented on charmstore/lib.py:
```diff
@@ -95,12 +87,14 @@ def __init__(self, id=None, api=None, timeout=None):
         self.series = None
         self.maintainer = None
         self.revision = None
+        self.revisions = None
```
This overwrites the method and will lead to `NoneType is not callable` errors.

(It's also less than ideal that this is one of the few attributes that is a method rather than a property or simple attribute, but I guess that's the existing interface, so it can't be changed.)

ChrisMacNaughton replied:
self.revisions is _not_ an existing field, it is a new field

johnsca commented on charmstore/lib.py:
```diff
@@ -115,10 +109,18 @@ def __init__(self, id=None, api=None, timeout=None):
 
     def revisions(self):
-        data = self.raw.get('revision-info', {}).get('Revisions', [])
+        if self.revisions is None:
```
As mentioned above, there's a conflict with this and the method name.  Something like `self._revisions` should be used as the caching variable instead.

johnsca replied:
These references still need to be fixed.

johnsca commented on charmstore/lib.py:
```diff
@@ -115,10 +109,18 @@ def __init__(self, id=None, api=None, timeout=None):
+        if self._files_fetched is False:
+            self.files = [
+                f.get('Name') for f in
+                    self.theblues._meta(self.id, ['manifest']) ]
```
I take it the response to this call doesn't include the contents?

ChrisMacNaughton replied:
the `_meta` call contains the actual HTTP request to the charmstore. What this line is doing is asking the charmstore for the "manifest" field of the given `Entity`

johnsca replied:
Yeah, sorry I was unclear.  I just wanted to confirm that the response from the `_meta` call didn't also contain the contents that we were throwing away here only to re-request subsequently.  Pretty sure that wouldn't be the case because that would be too much data, but you never know.

ChrisMacNaughton commented on charmstore/lib.py:
```diff
@@ -7,21 +7,13 @@
 
 
 AVAILABLE_INCLUDES = [
```
This was never all available options, this is some copy/paste out of [theblues](https://github.com/juju/theblues/blob/develop/theblues/charmstore.py#L121) ;-)

johnsca commented:
Oh, I also meant to ask... This wraps theblues, but why is the wrapper necessary rather than just using theblues directly?  Is it just to provide a nicer API?  I was also under the impression that theblues was basically unmaintained; if this is maintained and used, would it make more sense to fold the actual request logic into this as well?  If nothing else, the name is way better.

ChrisMacNaughton committed:
make placeholder variable internal

ChrisMacNaughton commented:
I'm very much +1 to updating libcharmstore to handle the charmstore API rather than using theblues

marcoceppi commented:
@johnsca At the time this library was created theblues didn't really have a good public interface. If theblues is unmaintained then we may want to consider folding request logic into this as well. We're still woefully young in the library's release cycle (0.0) so planning a 1.0 with request logic folded in and improvements to some of the less mature parts of the API would be great.

ChrisMacNaughton committed:
correct internal reference

ChrisMacNaughton committed:
change AVAILABLE_INCLUDES to DEFAULT_INCLUDES
